src/battle_ai_util.c

```diff
@@ -3249,82 +3283,55 @@ bool32 IsRecycleEncouragedItem(u32 item)
     return FALSE;
 }
 
-// score increases
-#define STAT_UP_2_STAGE     8
-#define STAT_UP_STAGE       10
 void IncreaseStatUpScore(u32 battlerAtk, u32 battlerDef, u32 statId, s32 *score)
 {
+    u32 noOfHitsToFaint = NoOfHitsForTargetToFaintAI(battlerDef, battlerAtk);
+    u32 aiIsFaster = GetWhichBattlerFaster(battlerAtk, battlerDef, TRUE) == AI_IS_FASTER;
+
     if (AI_DATA->abilities[battlerAtk] == ABILITY_CONTRARY)
         return;
 
-    if (AI_DATA->hpPercents[battlerAtk] < 80 && AI_RandLessThan(128))
+    // Don't increase stat if AI is at +4
+    if (gBattleMons[battlerAtk].statStages[statId] >= MAX_STAT_STAGE - 2)
         return;
 
-    if ((AI_THINKING_STRUCT->aiFlags[battlerAtk] & AI_FLAG_TRY_TO_FAINT) && CanAIFaintTarget(battlerAtk, battlerDef, 0))
-        return; // Damaging moves would get a score boost from AI_TryToFaint or PreferStrongestMove so we don't consider them here
+    // Don't increase stat if AI has less then 70% HP and number of hits isn't known
+    if (AI_DATA->hpPercents[battlerAtk] < 70 && noOfHitsToFaint == 0)
+        return;
 
     switch (statId)
     {
     case STAT_ATK:
-        if (HasMoveWithCategory(battlerAtk, BATTLE_CATEGORY_PHYSICAL) && AI_DATA->hpPercents[battlerAtk] > 40)
-        {
-            if (gBattleMons[battlerAtk].statStages[STAT_ATK] < STAT_UP_2_STAGE)
-                ADJUST_SCORE_PTR(2);
-            else if (gBattleMons[battlerAtk].statStages[STAT_ATK] < STAT_UP_STAGE)
-                ADJUST_SCORE_PTR(1);
-        }
-        if (HasMoveEffect(battlerAtk, EFFECT_FOUL_PLAY))
-            ADJUST_SCORE_PTR(1);
+        if ((HasMoveWithCategory(battlerAtk, BATTLE_CATEGORY_PHYSICAL))
+        && ((noOfHitsToFaint >= 2 && aiIsFaster) || (noOfHitsToFaint >= 3 && !aiIsFaster) || noOfHitsToFaint == 0))
```

Comment:
`(noOfHitsToFaint >= 2 && aiIsFaster) || (noOfHitsToFaint >= 3 && !aiIsFaster)` could be assigned to a `bool32` variable with a name? It seems to be used in a bunch of places. Not too sure what we'd call it, but I think logically this condition is "AI predicts that it can use a move next turn".

Non-blocking: I suppose in the future we'd be able to add a check for priority moves, e.g. `noOfHitsToFaint >= 2 && aiIsFaster` and the opponent's priority moves (if any) aren't enough to KO the AI.

EDIT: Also non-blocking, but I saw in the Evasion code that we have `!BattlerWillFaintFromWeather(...)`. That also seems like it could be part of whether the AI thinks it gets to use a move next turn.

Doubly-non-blocking: residual damage for the opponent (weather damage, poison, wrap, etc) could also influence the decision to use stat boosting/lowering effects. It's not worth lowering the stats of an opponent that's going to faint, and if we're slower and they're going to faint it's probably always worth going for a self-targeting move rather than wasted damage (EDIT: not true, they could switch).

Reply:
> (noOfHitsToFaint >= 2 && aiIsFaster) || (noOfHitsToFaint >= 3 && !aiIsFaster) could be assigned to a bool32 variable with a name? It seems to be used in a bunch of places. Not too sure what we'd call it, but I think logically this condition is "AI predicts that it can use a move next turn".

Yes!

> EDIT: Also non-blocking, but I saw in the Evasion code that we have !BattlerWillFaintFromWeather(...). That also seems like it could be part of whether the AI thinks it gets to use a move next turn.

This should be blocking. That's a good catch.

> Doubly-non-blocking: residual damage for the opponent (weather damage, poison, wrap, etc) could also influence the decision to use stat boosting/lowering effects. It's not worth lowering the stats of an opponent that's going to faint, and if we're slower and they're going to faint it's probably always worth going for a self-targeting move rather than wasted damage (EDIT: not true, they could switch).

I actually like that and I'm not sure if we should consider a switch since the AI can't know what the opposing side will switch into. It's probably better to just do damage in this case. 

> Non-blocking: I suppose in the future we'd be able to add a check for priority moves, e.g. noOfHitsToFaint >= 2 && aiIsFaster and the opponent's priority moves (if any) aren't enough to KO the AI.

I will try to keep that in mind but not sure how I would handle it optimally.

Reply:
Applied the first 3 points.